### DISCUSSION TITLE
How to identify which assembly file is being executed when cgemm is called from bench_blas.py?

### DISCUSSION THREAD

**Discussion by rajith-ravikumar-mcw** [OpenMathLib/OpenBLAS#5021] [Q&A]
There are several files located in the OpenBLAS/kernel/arm64/ directory that begin with the prefix "cgemm": 

- cgemm_kernel_4x4.S  
- cgemm_kernel_8x4.S  
- cgemm_kernel_8x4_cortexa53.c  
- cgemm_kernel_8x4_thunderx2t99.S  
- cgemm_kernel_sve_v1x4.S  
- cgemm_ncopy_sve_v1.c  
- cgemm_tcopy_sve_v1.c  

The script bench_blas.py calls functions for the input types c, d, s, and z, using input sizes of 100 and 1000.

I am looking to optimize these specific functions' implementations.

**Comment by martin-frbg** [comment]
The cpu on which the benchmark is being run determines which kernel will be used, through the CGEMMKERNEL entry in `kernel/arm64/KERNEL.target` (where target is one of the supported cpu names from TargetList.txt, like ARMV8, ARMV8SVE or CORTEXA57, etc). (These KERNEL files use Makefile syntax and are evaluated at build time)
If you are using a multi-cpu (DYNAMIC_ARCH=1) build, you can set the environment variable OPENBLAS_VERBOSE=2 to tell you
what cpu target was autodetected by the library
If the entry uses variable names CGEMM_UNROLL_M and/or CGEMM_UNROLL_N, you need to look up their values in the toplevel `param.h` first - this is where the "4x4" or "8x4" etc come from. There is no difference in CGEMM kernel choice between different sizes, but small sizes may be computed in single-threaded mode (see file interface/gemm.c for the threshold).

For SGEMM and DGEMM, there is now an additional set of "small" kernels that gets invoked directly, without passing through the block matrix setup code in driver/level3 first

**Comment by rajith-ravikumar-mcw** [comment]
Hi Martin,

Thanks for the clear explanation. I have added an example for reference to this thread.

For example:
Lets take cgemm test being executed on an armv8 machine

`OpenBLAS/kernel/arm64/KERNEL.ARMV8 ` contains the definition for `CGEMMKERNEL ` 

as

`CGEMMKERNEL    =  cgemm_kernel_$(CGEMM_UNROLL_M)x$(CGEMM_UNROLL_N).S`

The definitions for `$(CGEMM_UNROLL_M)` and `$(CGEMM_UNROLL_N)` are present in

`OpenBLAS/param.h`

For this example:
`/* Common ARMv8 parameters */
#if defined(ARMV8)
#define CGEMM_DEFAULT_UNROLL_M  8
#define CGEMM_DEFAULT_UNROLL_N  4`

So the kernel that is executed is
`OpenBLAS/kernel/arm64/cgemm_kernel_8x4.S`

**Comment by rajith-ravikumar-mcw** [comment]
Hi Martin,

Quick question: If I modify `cgemm_kernel_8x4.S`, build OpenBLAS, and use the new openblas.dll and openblas.lib to generate the pyd file with Meson for benchmarking , will the updated kernel be used?

**Reply by martin-frbg** [reply]
I would expect so - but to be sure, please do a complete rebuild (i.e. run  `make clean` or equivalent between builds) as the build file structure is a bit complicated and just rebuilding/replacing a single object may not always work as expected.

**Reply by MugundanMCW** [reply]
@martin-frbg I could see a tool [codespeed.io](https://codspeed.io/) that could provide execution details upto kernal execution, Is there any equivalent tool that could to the same as the tool I mentioned?
Thanks

**Reply by martin-frbg** [reply]
codspeed.io is what we use for the benchmark that runs on every PR, but I am almost certain that this service runs on Linux where one can easily use `perf` to get a performance analysis down to function level (and individual lines in the machine code of each). I do not know of any Windows tool that would do the same
